rust/xaynet-analytics/src/data_combination/data_combiner.rs

```diff
@@ -0,0 +1,178 @@
+            DataPointMetadata::new(Period::new(PeriodUnit::Weeks, 6), end_period),
+            DataPointMetadata::new(Period::new(PeriodUnit::Months, 3), end_period),
+        ]
+        .iter()
+        .map(|metadata| {
+            let period_thresholds = (0..metadata.period.n)
+                .map(|i| self.get_start_of_period(*metadata, i))
+                .collect();
+            CalcWasActiveEachPastPeriod::new(
+                *metadata,
+                self.filter_events_in_this_period(*metadata, self.get_all_events()),
+                period_thresholds,
+            )
+        })
+        .map(DataPoint::WasActiveEachPastPeriod)
+        .collect()
+    }
+
+    // TODO: return an iterator instead of Vec: https://xainag.atlassian.net/browse/XN-1517
+    fn get_all_events(&self) -> Vec<AnalyticsEvent> {
+        self.repo.get_all_events()
+    }
+
+    /// TODO: don't use String here, handle via RouteController
+    fn get_all_screen_routes(&self) -> Vec<String> {
+        self.repo.get_all_routes()
+    }
+
+    // TODO: return an iterator instead of Vec: https://xainag.atlassian.net/browse/XN-1517
+    fn filter_events_in_this_period(
+        &self,
+        metadata: DataPointMetadata,
+        events: Vec<AnalyticsEvent>,
+    ) -> Vec<AnalyticsEvent> {
+        let start_of_period = self.get_start_of_period(metadata, metadata.period.n);
+        self.filter_events_before_end_of_period(metadata.end, events)
+            .into_iter()
+            .filter(|event| event.timestamp > start_of_period)
+            .collect()
+    }
+
+    fn get_start_of_period(&self, metadata: DataPointMetadata, n_periods: u32) -> DateTime<Utc> {
+        let avg_days_per_month = 365.0 / 12.0;
+        let midnight_end_of_period = get_midnight(metadata.end);
+        match metadata.period.unit {
+            PeriodUnit::Days => midnight_end_of_period - Duration::days(n_periods as i64),
+            PeriodUnit::Weeks => midnight_end_of_period - Duration::weeks(n_periods as i64),
+            PeriodUnit::Months => {
+                midnight_end_of_period
+                    - Duration::days((n_periods as f64 * avg_days_per_month) as i64)
+            }
+        }
+    }
+
+    // TODO: return an iterator instead of Vec: https://xainag.atlassian.net/browse/XN-1517
+    fn filter_events_before_end_of_period(
+        &self,
+        end_of_period: DateTime<Utc>,
+        events: Vec<AnalyticsEvent>,
+    ) -> Vec<AnalyticsEvent> {
+        let midnight_end_of_period = get_midnight(end_of_period);
+        events
+            .into_iter()
+            .filter(|event| event.timestamp < midnight_end_of_period)
+            .collect()
+    }
+
+    // TODO: return an iterator instead of Vec: https://xainag.atlassian.net/browse/XN-1517
+    fn get_events_single_route(&self, route: String) -> Vec<AnalyticsEvent> {
+        self.get_all_events()
+            .into_iter()
+            .filter(|event| event.screen_route.as_ref().unwrap() == &route)
```

Comment:
Is calling `unwrap` safe here (is `screen_route` always `Some`)?
because if screen_route is `None`, `unwarp` will `panic` (the program will crash).

you can write something like this (if it is not guaranteed that `screen_route` is always `Some`)

```rust
        self.get_all_events()
            .into_iter()
            .filter(|event| {
                if let Some(ref screen_route) = event.screen_route {
                    screen_route == &route
                } else {
                    false
                }
            })
            .collect()
```

Reply:
thanks, i'll add it. still using stuff without understanding 100%..